### DISCUSSION TITLE
.Net Standard 2.0 Limitation

### DISCUSSION THREAD

**Discussion by cpsusie** [dotnet/roslyn#56532] [Q&A]
When will we be able to make analyzers and source generators using .NET 5.0 or .NET 6?  How long will we be limited to using .NET Standard 2.0?  I would really like to be able to use C# 9 and C# 10 features when writing analyzers and source generators.

**Comment by PathogenDavid** [comment]
> How long will we be limited to using .NET Standard 2.0?

[Not for the foreseeable future.](https://github.com/dotnet/roslyn/issues/47087#issuecomment-901429725)

The .NET Standard 2.0 limitation is due to Visual Studio using .NET Framework. Both analyzers and source generators need to be able to run within Visual Studio to support the IDE experience.

--------------

> I would really like to be able to use C# 9 and C# 10 features when writing analyzers and source generators.

It's not officially supported, but as long as they don't require runtime support you can use C# 9 and C# 10 features on older runtimes  by manually specifying `LangVersion` in your csproj. Some features require you to add a little boilerplate to provide types the compiler needs to use them. (Roslyn will typically tell you if you try to use a feature you can't or if you need to manually provide a type.)

Here's a list of what works and what doesn't for C# 9 https://github.com/dotnet/roslyn/discussions/47701#discussioncomment-79566 and for C# 10: https://github.com/dotnet/roslyn/discussions/47701#discussioncomment-1356495
